agent/app/agent_windows.go

```diff
@@ -0,0 +1,188 @@
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package app
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/engine"
+	"github.com/aws/amazon-ecs-agent/agent/sighandlers"
+	"github.com/aws/amazon-ecs-agent/agent/statemanager"
+	"github.com/cihub/seelog"
+	"golang.org/x/sys/windows/svc"
+)
+
+const (
+	//EcsSvcName is the name of the service
+	EcsSvcName = "AmazonECS"
+)
+
+// startWindowsService runs the ECS agent as a Windows Service
+func (agent *ecsAgent) startWindowsService() int {
+	svc.Run(EcsSvcName, newHandler(agent))
+	return 0
+}
+
+// handler implements https://godoc.org/golang.org/x/sys/windows/svc#Handler
+type handler struct {
+	ecsAgent agent
+}
+
+func newHandler(agent agent) *handler {
+	return &handler{agent}
+}
+
+// Execute implements https://godoc.org/golang.org/x/sys/windows/svc#Handler
+// The basic way that this implementation works is through two channels (representing the requests from Windows and the
+// responses we're sending to Windows) and two goroutines (one for message processing with Windows and the other to
+// actually run the agent).  Once we've set everything up and started both goroutines, we wait for either one to exit
+// (the Windows goroutine will exit based on messages from Windows while the agent goroutine exits if the agent exits)
+// and then cancel the other.  Once everything has stopped running, this function returns and the process exits.
+func (h *handler) Execute(args []string, requests <-chan svc.ChangeRequest, responses chan<- svc.Status) (bool, uint32) {
+	// channels for communication between goroutines
+	ctx, cancel := context.WithCancel(context.Background())
+	agentDone := make(chan struct{})
+	windowsDone := make(chan struct{})
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+
+	go func() {
+		defer close(windowsDone)
+		defer wg.Done()
+		h.handleWindowsRequests(ctx, requests, responses)
+	}()
+
+	go func() {
+		defer close(agentDone)
+		defer wg.Done()
+		h.runAgent(ctx)
+	}()
+
+	// Wait until one of the goroutines is either told to stop or fails spectacularly.  Under normal conditions we will
+	// be waiting here for a long time.
+	select {
+	case <-windowsDone:
+		// Service was told to stop by the Windows API.  This happens either through manual intervention (i.e.,
+		// "Stop-Service ECS") or through system shutdown.  Regardless, this is a normal exit event and not an error.
+		seelog.Info("Received normal signal from Windows to exit")
+	case <-agentDone:
+		// This means that the agent stopped on its own.  This is where it's appropriate to light the event log on fire
+		// and set off all the alarms.
+		seelog.Error("Exiting")
```

Comment:
Can we add more context in this exit log message?

Reply:
There isn't a ton of context here except that the agent stopped on its own... What would you like me to add here?